docs/docs/concepts/triggers/KeySpace_Triggers.md

```diff
@@ -0,0 +1,205 @@
+          13) "last_exection_time"
+          14) (integer) 0
+          15) "total_exection_time"
+          16) (integer) 0
+          17) "avg_exection_time"
+          18) "0"
+```
+
+## Triggers Guarantees
+
+If the callback pass to the trigger is a `JS` function (not a Coroutine), it is guarantee that the callback will be invoke atomically along side the operation that cause the trigger, i.e all client will see the data only after the callback has finished. In addition, it is guarantee that the effect of the callback will be replicated to the replication and the AOF in a `multi/exec` block together with the command that fired the trigger.
+
+If the callback is a Coroutine, it will be executed in the background and there is not guarantees on where or if it will be executed. The guarantees are the same as describe on [Sync and Async invocation](../Sync_Async.md).
+
+## Upgrades
+
+When upgrading the trigger code (using the `REPLACE` option of `TFUNCTION LOAD` command) all the trigger parameters can be modified.
+
+## Advanced Usage
+
+For most use cases, `registerKeySpaceTrigger` API is enough. But there are some use cases where you might need a better guaranteed on when the trigger will be fired. Lets look at the following example:
+
+```js
+#!js api_version=1.0 name=myFirstLibrary
+
+redis.registerKeySpaceTrigger("consumer", "", function(client, data){
+    if (client.call("type", data.key) != "hash") {
+        // key is not a has, do not touch it.
+        return;
+    }
+    var name = client.call('hget', data.key, 'name');
+    client.call('incr', `name_${name}`);
+});
+```
+
+Whenever a hash key is changing, the example above will read the field `name` from the hash (assume its `tom`) and increase the value of the key `name_tom` (**notice that this function will not work properly on cluster, we need to use `{}` on the keys name to make sure we are writing to a key located on the current shard, for simplicity we ignore the cluster issues now**). Running the function will give the following results:
+
+```bash
+127.0.0.1:6379> hset x name tom
+(integer) 1
+127.0.0.1:6379> hset x name jerry
+(integer) 0
+127.0.0.1:6379> get name_tom
+"1"
+127.0.0.1:6379> get name_jerry
+"1"
+```
+
+We can see that the key `name_tom` was increased once, and the key `name_jerry` was increased once. Will we get the same results if we wrap the `hset`'s with a `multi`/`exec`?
+
+```bash
+127.0.0.1:6379> multi
+OK
+127.0.0.1:6379(TX)> hset x name tom
+QUEUED
+127.0.0.1:6379(TX)> hset x name jerry
+QUEUED
+127.0.0.1:6379(TX)> exec
+1) (integer) 1
+2) (integer) 0
+127.0.0.1:6379> get name_tom
+(nil)
+127.0.0.1:6379> get name_jerry
+"2"
+```
+
+What just happened? `name_jerry` was increased twice while `name_tom` was not increased at all. This happened because in case of a `multi`/`exec` or Lua, the notifications are fire at the end of the transaction, so all the notifications will see the last value that was written which is `jerry`.
```

Comment:
```suggestion
What just happened? `name_jerry` was incremented twice while `name_tom` was not incremented at all. This happened because, in case of a `multi`/`exec` or Lua function, the notifications are fired at the end of the transaction, so all the clients will receive notifications of the last value written, which is `jerry`.
```